lib/livebook/web_socket/client.ex

```diff
@@ -0,0 +1,252 @@
+defmodule Livebook.WebSocket.Client do
+  @moduledoc false
+  use GenServer
+
+  require Logger
+
+  defstruct [
+    :conn,
+    :request_ref,
+    :websocket,
+    :caller,
+    :status,
+    :resp_headers,
+    :resp_body,
+    :ref,
+    closing?: false
+  ]
+
+  defmodule Error do
+    defstruct [:body, :status, :headers]
+  end
+
+  @ws_path "/livebook/websocket"
+
+  def start(opts \\ []) do
+    GenServer.start(__MODULE__, opts)
+  end
+
+  def start_link(opts \\ []) do
+    GenServer.start_link(__MODULE__, opts)
+  end
+
+  @doc """
+  Connects the WebSocket client.
+  """
+  def connect(socket, url, headers \\ []) do
+    GenServer.call(socket, {:connect, url, headers})
+  end
+
+  @doc """
+  Closes the socket
+  """
+  def close(socket) do
+    GenServer.cast(socket, :close)
+  end
+
+  @doc """
+  Sends a message to the WebSocket server the message request.
+  """
+  def send_message(socket, data) when is_struct(data) do
+    send_message(socket, Protobuf.encode(data))
+  end
+
+  def send_message(socket, data) when is_binary(data) do
+    GenServer.call(socket, {:send_message, data})
+  end
+
+  ## GenServer callbacks
+
+  @doc false
+  def init(_) do
+    {:ok, %__MODULE__{}}
+  end
+
+  @dialyzer {:nowarn_function, handle_call: 3}
+
+  @doc false
+  def handle_call({:connect, url, headers}, from, state) do
+    uri = URI.parse(url)
+
+    http_scheme = parse_http_scheme(uri)
+    ws_scheme = parse_ws_scheme(uri)
+    path = parse_ws_path(uri)
+
+    with {:ok, conn} <- Mint.HTTP.connect(http_scheme, uri.host, uri.port),
+         {:ok, conn, ref} <- Mint.WebSocket.upgrade(ws_scheme, conn, path, headers) do
+      state = %{state | conn: conn, request_ref: ref, caller: from}
+      {:noreply, state}
+    else
+      {:error, reason} ->
+        {:reply, {:error, reason}, state}
+
+      {:error, conn, reason} ->
+        {:reply, {:error, reason}, %{state | conn: conn}}
+    end
+  end
+
+  def handle_call({:send_message, msg}, _from, state) do
+    case stream_frame(state, {:text, msg}) do
+      {:ok, state} -> {:reply, :ok, state}
+      {:error, state, reason} -> {:reply, {:error, reason}, state}
+    end
+  end
+
+  @doc false
+  def handle_cast(:close, state) do
+    do_close(state)
+  end
+
+  @doc false
+  def handle_info(message, state) do
+    case Mint.WebSocket.stream(state.conn, message) do
```

Comment:
I think we could simplify this by doing the following:

1. Try to establish a connection and do the upgrade outside of the process
2. Once the connection is upgraded, create the process, and transfer the socket ownership to the GenServer

WDYT?